distributed/scheduler.py

```diff
@@ -8570,9 +8601,17 @@ def _materialize_graph(
             deps.remove(k)
         dependencies[k] = deps
 
-    # Remove aliases
+    # FIXME: Remove this section. This condition is generated by
+    # unpack_remotedata when using persisted collections. There should be a
+    # better way to achieve that tasks are not self referencing themselves
+
+    # Remove aliases.
     for k in list(dsk):
-        if dsk[k] is k:
+        # Most keys are tuples so the `is` comparison does not work
+        # However, we cannot simply use `==` since dsk[k] may contain actual
+        # payload data, e.g. a pandas DataFrame where equals is not necessarily
+        # properly defined with any object
+        if dsk[k] is k or tokenize(dsk[k]) == tokenize(k):
```

Comment:
Again, isn't this very expensive in case of bulky embedded constants?
What's the difference from `type(dsk[k]) is type(k) and dsk[k] == k` (which would be drastically faster?)

Reply:
Actually above there's an edge case where k is a tuple and dsk[k] is a tuple of numpy arrays. Can be fixed by a utility function `is_same_key`.

Reply:
`iskey(a) and iskey(b) and a == b``?